File: structural/dart/facade_structured/facades.dart
import 'subsystems.dart';

class ComputerFacade {
  final CPU _cpu = CPU();
  final Memory _memory = Memory();
  final HardDrive _hardDrive = HardDrive();
  
  void start() {
    print('=== Starting Computer ===');
    _cpu.freeze();
    _memory.load(0, _hardDrive.read(0, 1024));
    _cpu.jump(0);
    _cpu.execute();
    print('=== Computer Started ===');
  }
  
  void shutdown() {
    print('=== Shutting Down Computer ===');
    print('CPU: Stopping...');
    print('Memory: Clearing...');
    print('HardDrive: Spinning down...');
    print('=== Computer Shutdown ===');
  }
}

class HomeTheaterFacade {
  final Amplifier _amp = Amplifier();
  final Tuner _tuner = Tuner();
  final DVDPlayer _dvd = DVDPlayer();
  final Projector _projector = Projector();
  final TheaterLights _lights = TheaterLights();
  final ProjectionScreen _screen = ProjectionScreen();
  final PopcornPopper _popper = PopcornPopper();
  
  void watchMovie(String movie) {
    print('\n=== Get ready to watch a movie ===');
    _popper.on(); _popper.pop();
    _lights.dim(10);
    _screen.down();
    _projector.on(); _projector.setInput('DVD');
    _amp.on(); _amp.setVolume(5);
    _dvd.on(); _dvd.play(movie);
    print('=== Movie is playing ===');
  }
  
  void endMovie() {
    print('\n=== Shutting movie theater down ===');
    _popper.off();
    _lights.on();
    _screen.up();
    _projector.off();
    _amp.off();
    _dvd.off();
    print('=== Movie theater shut down ===');
  }
  
  void listenToRadio(double frequency) {
    print('\n=== Tuning in to radio ===');
    _amp.on(); _amp.setVolume(5);
    _tuner.on(); _tuner.setFrequency(frequency);
    print('=== Radio is playing ===');
  }
  
  void turnOffRadio() {
    print('\n=== Shutting down radio ===');
    _tuner.off();
    _amp.off();
    print('=== Radio turned off ===');
  }
}

class OrderFacade {
  final InventorySystem _inventory = InventorySystem();
  final PaymentSystem _payment = PaymentSystem();
  final ShippingSystem _shipping = ShippingSystem();
  final NotificationSystem _notification = NotificationSystem();
  
  bool placeOrder(String item, int quantity, double amount, String address, String email) {
    print('\n=== Processing Order ===');
    if (!_inventory.checkInventory(item, quantity)) {
      print('Order failed: Item not in stock');
      return false;
    }
    if (!_payment.processPayment(amount)) {
      print('Order failed: Payment declined');
      return false;
    }
    _inventory.updateInventory(item, -quantity);
    _shipping.scheduleShipping(address);
    _notification.sendConfirmation(email);
    print('=== Order Placed Successfully ===');
    return true;
  }
  
  void cancelOrder(String item, int quantity, double amount, String email) {
    print('\n=== Cancelling Order ===');
    _inventory.updateInventory(item, quantity);
    _payment.refundPayment(amount);
    _shipping.cancelShipping();
    _notification.sendCancellation(email);
    print('=== Order Cancelled ===');
  }
}

File: structural/typescript/facade_structured/facades.ts
import { CPU, Memory, HardDrive, Amplifier, Tuner, DVDPlayer, Projector, TheaterLights, Screen, PopcornPopper, InventorySystem, PaymentSystem, ShippingSystem, NotificationSystem } from './subsystems';

export class ComputerFacade {
  private cpu: CPU = new CPU();
  private memory: Memory = new Memory();
  private hardDrive: HardDrive = new HardDrive();
  
  start(): void {
    console.log('=== Starting Computer ===');
    this.cpu.freeze();
    this.memory.load(0, this.hardDrive.read(0, 1024));
    this.cpu.jump(0);
    this.cpu.execute();
    console.log('=== Computer Started ===');
  }
  shutdown(): void {
    console.log('=== Shutting Down Computer ===');
    console.log('CPU: Stopping...');
    console.log('Memory: Clearing...');
    console.log('HardDrive: Spinning down...');
    console.log('=== Computer Shutdown ===');
  }
}

export class HomeTheaterFacade {
  private amp: Amplifier = new Amplifier();
  private tuner: Tuner = new Tuner();
  private dvd: DVDPlayer = new DVDPlayer();
  private projector: Projector = new Projector();
  private lights: TheaterLights = new TheaterLights();
  private screen: Screen = new Screen();
  private popper: PopcornPopper = new PopcornPopper();
  
  watchMovie(movie: string): void {
    console.log('\n=== Get ready to watch a movie ===');
    this.popper.on(); this.popper.pop();
    this.lights.dim(10);
    this.screen.down();
    this.projector.on(); this.projector.setInput('DVD');
    this.amp.on(); this.amp.setVolume(5);
    this.dvd.on(); this.dvd.play(movie);
    console.log('=== Movie is playing ===');
  }
  endMovie(): void {
    console.log('\n=== Shutting movie theater down ===');
    this.popper.off(); this.lights.on(); this.screen.up();
    this.projector.off(); this.amp.off(); this.dvd.off();
    console.log('=== Movie theater shut down ===');
  }
  listenToRadio(frequency: number): void {
    console.log('\n=== Tuning in to radio ===');
    this.amp.on(); this.amp.setVolume(5);
    this.tuner.on(); this.tuner.setFrequency(frequency);
    console.log('=== Radio is playing ===');
  }
  turnOffRadio(): void {
    console.log('\n=== Shutting down radio ===');
    this.tuner.off(); this.amp.off();
    console.log('=== Radio turned off ===');
  }
}

export class OrderFacade {
  private inventory: InventorySystem = new InventorySystem();
  private payment: PaymentSystem = new PaymentSystem();
  private shipping: ShippingSystem = new ShippingSystem();
  private notification: NotificationSystem = new NotificationSystem();
  
  placeOrder(item: string, quantity: number, amount: number, address: string, email: string): boolean {
    console.log('\n=== Processing Order ===');
    if (!this.inventory.checkInventory(item, quantity)) {
      console.log('Order failed: Item not in stock');
      return false;
    }
    if (!this.payment.processPayment(amount)) {
      console.log('Order failed: Payment declined');
      return false;
    }
    this.inventory.updateInventory(item, -quantity);
    this.shipping.scheduleShipping(address);
    this.notification.sendConfirmation(email);
    console.log('=== Order Placed Successfully ===');
    return true;
  }
  cancelOrder(item: string, quantity: number, amount: number, email: string): void {
    console.log('\n=== Cancelling Order ===');
    this.inventory.updateInventory(item, quantity);
    this.payment.refundPayment(amount);
    this.shipping.cancelShipping();
    this.notification.sendCancellation(email);
    console.log('=== Order Cancelled ===');
  }
}
